File: android/feature/dua/src/main/kotlin/com/fatwabot/feature/dua/DuaModels.kt
package com.fatwabot.feature.dua

import java.io.File
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/** Local-only favorite record — mirror of iOS FavoriteDua (docs/features/dua.md,
 * no backend sync in M2). */
@Serializable
data class FavoriteDua(val duaId: String, val addedAtEpochSeconds: Long)

interface DuaStoring {
    fun loadFavorites(): List<FavoriteDua>
    fun saveFavorites(favorites: List<FavoriteDua>)
}

class FileDuaStore(private val file: File) : DuaStoring {
    private val json = Json { ignoreUnknownKeys = true }
    private val serializer = ListSerializer(FavoriteDua.serializer())

    override fun loadFavorites(): List<FavoriteDua> = runCatching {
        if (!file.exists()) emptyList() else json.decodeFromString(serializer, file.readText())
    }.getOrDefault(emptyList())

    override fun saveFavorites(favorites: List<FavoriteDua>) {
        runCatching {
            file.parentFile?.mkdirs()
            val tmp = File(file.parentFile, "${file.name}.tmp")
            tmp.writeText(json.encodeToString(serializer, favorites))
            tmp.renameTo(file)
        }
    }
}

/**
 * Arabic-aware search normalization — mirror of iOS DuaSearch
 * (docs/features/dua.md: "diacritic-insensitive for Arabic — strip tashkeel
 * before matching"). Uses explicit \\uXXXX code-point escapes (not literal
 * combining marks, unreliable to author/verify inline) for the same ranges
 * as iOS: U+0610-061A, U+064B-065F, U+0670, U+06D6-06DC, U+06DF-06E8,
 * U+06EA-06ED (Arabic tashkeel / Quranic annotation marks).
 */
object DuaSearch {
    private val tashkeel = Regex(
        "[ؐ-ًؚ-ٰٟۖ-ۜ۟-۪ۨ-ۭ]",
    )

    fun normalize(text: String): String = text.replace(tashkeel, "").lowercase().trim()
}

File: ios/Packages/AppFeatures/Sources/DuaFeature/DuaModels.swift
import Foundation

/// Local-only favorite record (docs/features/dua.md — no backend sync in M2).
public struct FavoriteDua: Codable, Equatable, Sendable {
    public let duaId: String
    public let addedAt: Date

    public init(duaId: String, addedAt: Date) {
        self.duaId = duaId
        self.addedAt = addedAt
    }
}

public protocol DuaStoring: Sendable {
    func loadFavorites() -> [FavoriteDua]
    func saveFavorites(_ favorites: [FavoriteDua])
}

public struct FileDuaStore: DuaStoring {
    private let fileURL: URL

    public init(directory: URL) {
        fileURL = directory.appendingPathComponent("dua-favorites.json")
    }

    public func loadFavorites() -> [FavoriteDua] {
        guard let data = try? Data(contentsOf: fileURL) else { return [] }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return (try? decoder.decode([FavoriteDua].self, from: data)) ?? []
    }

    public func saveFavorites(_ favorites: [FavoriteDua]) {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(favorites) else { return }
        try? FileManager.default.createDirectory(
            at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try? data.write(to: fileURL, options: .atomic)
    }
}

/// Arabic-aware search normalization (docs/features/dua.md: "diacritic-insensitive
/// for Arabic — strip tashkeel before matching").
enum DuaSearch {
    private static let tashkeel = CharacterSet(charactersIn: "\u{0610}-\u{061A}\u{064B}-\u{065F}\u{0670}\u{06D6}-\u{06DC}\u{06DF}-\u{06E8}\u{06EA}-\u{06ED}")

    static func normalize(_ text: String) -> String {
        String(String.UnicodeScalarView(text.unicodeScalars.filter { !tashkeel.contains($0) }))
            .lowercased()
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
